crates/yttrium/src/chain_abstraction/client.rs

```diff
@@ -49,6 +71,213 @@ impl Client {
         }
     }
 
+    pub async fn get_route_ui_fields(
+        &self,
+        route_response: RouteResponseAvailable,
+        initial_transaction: Transaction,
+        currency: Currency,
+        // TODO use this to e.g. modify priority fee
+        _speed: String,
+    ) -> Result<RouteUiFields, RouteUiFieldsError> {
+        let chains = route_response
+            .transactions
+            .iter()
+            .chain(std::iter::once(&initial_transaction))
+            .map(|t| t.chain_id.clone())
+            .collect::<HashSet<_>>();
+        println!("chains: {chains:?}");
+
+        let mut providers = HashMap::new();
+        for chain_id in &chains {
+            let mut url = self.base_url.join(PROXY_ENDPOINT_PATH).unwrap();
+            url.query_pairs_mut()
+                .append_pair("chainId", chain_id)
+                .append_pair("projectId", self.project_id.as_ref());
+            let provider = ReqwestProvider::<Ethereum>::new(RpcClient::new(
+                Http::with_client(self.client.clone(), url),
+                false,
+            ));
+            providers.insert(chain_id, provider);
+        }
+
+        // TODO run fungible lookup, eip1559_fees, and l1 data fee, in parallel
+
+        let addresses = chains
+            .iter()
+            .map(|t| format!("{}:{}", t, NATIVE_TOKEN_ADDRESS))
+            .collect();
+        println!("addresses: {addresses:?}");
+        let response = self
+            .client
+            .post(self.base_url.join(FUNGIBLE_PRICE_ENDPOINT_PATH).unwrap())
+            .json(&PriceRequestBody {
+                project_id: self.project_id.clone(),
+                currency,
+                addresses,
+            })
+            .send()
+            .await
+            .map_err(RouteUiFieldsError::Request)?;
+        let prices = if response.status().is_success() {
+            response
+                .json::<PriceResponseBody>()
+                .await
+                .map_err(RouteUiFieldsError::Json)
+        } else {
+            Err(RouteUiFieldsError::RequestFailed(
+                response.status(),
+                response.text().await,
+            ))
+        }?;
+
+        let mut eip1559_fees = HashMap::new();
+        for chain_id in &chains {
+            let estimate = providers
+                .get(chain_id)
+                .unwrap()
+                .estimate_eip1559_fees(None)
+                .await
+                .unwrap();
+            eip1559_fees.insert(chain_id, estimate);
+        }
+
+        async fn estimate_gas_fees(
```

Comment:
Curious: Here you're defining a function inside a function, right? Kotlin allows for the same, cool

Reply:
Yes indeed!